File: Python/08.py
"""

Project Euler Solutions - Problem 08

By Chris Mander (http://github.com/manderc3)

"""

import time

def getNextChar(f):
	""" Returns next char in file that isn't a newline. """
	char = f.read(1)
	while char == "\n":
		char = f.read(1)
	return char

def adjProduct(f_path, m_range):
	"""
	Scans a file and returns the highest product from a arbitrary range of ints.
	This function is strictly for files that only contain integers.
	"""
	f = open(f_path, "r")

	h_prd = 0

	subjects = ["" for i in range(m_range)]
	
	for i in range(len(subjects)):
		subjects[i] = getNextChar(f)

	while subjects[len(subjects) - 1]:
		c_prd = 1

		for i in subjects:
			c_prd *= int(i)

		if c_prd > h_prd:
			h_prd = c_prd

		char = getNextChar(f)

		subjects = subjects[1:] + [char]

	f.close()

	return h_prd

if __name__ == "__main__":
	start = time.clock()
	print(adjProduct("files/1000digits.txt", 13))
	print("Time: " + str(time.clock() - start))

File: C/08.c
#include <stdio.h>
#include <stdlib.h>
#include "common.h"

size_t adjProduct(char*, const int);
void updateSubjects(int*, char*, const int);

void updateSubjects(int* arr, char* c, int arr_size) {
	int i;

	for (i = 1; i < arr_size; i++) {
		arr[i - 1] = arr[i];		
	}

	arr[arr_size - 1] = ((int) *c) - 48;
}

size_t adjProduct(char* f_path, int m_range) {
	FILE* file = fopen(f_path, "r");
	size_t c_prd, h_prd = 0;
	char c;

	if (file == NULL) {
		exit(-1);
	}

	int* subjects = malloc(sizeof(int) * m_range);
	int i;

	for (i = 0; i < m_range; i++)
		subjects[i] = ((int) getNextChar(file)) - 48;

	do {
		c_prd = 1;

		for (i = 0; i < m_range; i++)
			c_prd *= subjects[i];

		if (c_prd > h_prd)
			h_prd = c_prd;

		c = getNextChar(file);
		updateSubjects(subjects, &c, m_range);
	} while (c != EOF);

	free(subjects);
	return h_prd;
}

int main() {

	printf("%zu\n", adjProduct("./files/1000digits.txt", 13));
	return 0;
}